src/Orleans/Runtime/RequestContext.cs

```diff
@@ -102,38 +115,30 @@ public static bool Remove(string key)
 
         public static void Import(Dictionary<string, object> contextData)
         {
+#if !NETSTANDARD
             if (PropagateActivityId)
             {
                 object activityIdObj = Guid.Empty;
-                if (contextData?.TryGetValue(E2_E_TRACING_ACTIVITY_ID_HEADER, out activityIdObj) != true)
+                if (contextData?.TryGetValue(E2_E_TRACING_ACTIVITY_ID_HEADER, out activityIdObj) == true)
                 {
-                    if (ActivityId.Value != Guid.Empty)
-                    {
-                        ActivityId.Value = Guid.Empty;
-                    }
-
-#if !NETSTANDARD
-                    object legacyActivityIdObj = null;
-                    if (contextData?.TryGetValue(E2_E_TRACING_LEGACY_ACTIVITY_ID_HEADER, out legacyActivityIdObj) == true)
-                    {
-                        Trace.CorrelationManager.ActivityId = (Guid) legacyActivityIdObj;
-                    }
-#endif
+                    Trace.CorrelationManager.ActivityId = (Guid)activityIdObj;
```

Comment:
but that is what we were doing all along anyway, since `Trace.CorrelationManager.ActivityId` has special meaning in .NET.
I don't think there's much we can do. Nevertheless, I did optimize the Export, so that it doesn't modify the current CallContextData by adding that to the dictionary. It is only added to the dictionary being returned, so that it's serialized in the grain call.